### ISSUE TITLE
ADAM can't install APK on some devices

### ISSUE THREAD

**Issue by snpefk** [MarathonLabs/marathon#436]
**Describe the bug**
Trying new `ADAM` provider I start a usual test run on several devices. Marathon wasn't able to install apk on all devices although response from device was `Success`. Here sample from log:
```log
Success
 [0;39mcom.malinskiy.marathon.android.exception.InstallException: pkg: /data/local/tmp/debug.apk
	at com.malinskiy.marathon.android.AndroidAppInstaller$reinstall$2.invokeSuspend(AndroidAppInstaller.kt:49)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:32)
	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:113)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**To Reproduce**
Steps to reproduce the behaviour:
1. Configuration of Marathonfile
```yaml
vendorConfiguration:
  androidSdk: "/opt/android/sdk"
  applicationApk: /some/path/to/debug/debug.apk
  testApplicationApk: /some/path/to/debug/debug-AndroidTest.apk
  vendor: ADAM
  type: "Android"
  autoGrantPermission: false
  applicationPmClear: true
  testApplicationPmClear: true
  serialStrategy: "ddms"
  screenRecordConfiguration:
    preferableRecorderType: "VIDEO"
name: "Tests"
outputDir: "./marathon"
filteringConfiguration:
  allowlist:
  - type: "composition"
    filters:
    - type: "package"
      regex: "heh\\.mda\\.feature.*"
    op: "UNION"
  blocklist: []
testOutputTimeoutMillis: 150000
debug: true
ignoreFailures: true
retryStrategy:
  type: "fixed-quota"
  totalAllowedRetryQuota: 65
  retryPerTestQuota: 1
```
2. Start marathon 

**Expected behavior**
Adam install apk on emulator 

**Logs and reports**
[marathon-adam.log](https://github.com/Malinskiy/marathon/files/5871849/marathon-adam.log)

**Devices (please complete the following information):**
 - Device: emulator
 - OS: Android API 23

**Additional context**

**Comment by mikepenz** [comment]
@snpefk I had similar error logs, the problem is that `adam` (in 0.6.0) requires the phone to be connected with **wireless adb** instead of USB (not sure how this works with emulators but I would assume it behaves the same) 

There's an open PR which is meant to resolve this: https://github.com/Malinskiy/marathon/pull/426

**Comment by Malinskiy** [comment]
This issue is related to unexpected modification of firmware (or fragmentation, whatever you want to call it). The installation logic by default checks the output of the installation command and if it starts with `Success` then it considers the installation successful. ADB shell doesn't provide the exit code by default (only on new devices), so we have to resort to parsing the output.
In your case the installation message is 
```
pkg: /data/local/tmp/qpay-develop-gms-debug-3.62.0.apk
Success
```
I've seen this problem and there is already a fix in the branch for Adam stabilisation mentioned above by @mikepenz. The fix is now just to check if the output contains the word Success. This is ofc prone to error: if your APK filename contains `Success` this whole thing just fails. Ddmlib is no different in this case though. I'm considering if we should just check the exit code via a hack I did with Adam with emulating the exit code instead of parsing all the possible ways how firmware creators could modify the output of Am.java.

Anyways, should be fixed in 0.6.1

**Comment by snpefk** [comment]
Thanks for a detailed answer. 
I'll leave this issue open until release of Marathon 0.6.1 and retesting this behavior.
